client.go

```diff
@@ -0,0 +1,788 @@
+package redis
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/dop251/goja"
+	"github.com/go-redis/redis/v8"
+	"go.k6.io/k6/js/modules"
+)
+
+// Client represents the Client constructor (i.e. `new redis.Client()`) and
+// returns a new Redis client object.
+type Client struct {
+	vu          modules.VU
+	redisClient redis.UniversalClient
+}
+
+// Set the given key with the given value.
+func (c *Client) Set(key string, value interface{}, expiration int) *goja.Promise {
+	promise, resolve, reject := c.makeHandledPromise()
+
+	if err := c.isSupportedType(value); err != nil {
+		reject(err)
+		return promise
+	}
+
+	go func() {
+		result, err := c.redisClient.Set(c.vu.Context(), key, value, time.Duration(expiration)*time.Second).Result()
+		if err != nil {
+			reject(err)
+		}
+
+		resolve(result)
+	}()
+
+	return promise
+}
+
+// Get returns the value for the given key.
+func (c *Client) Get(key string) *goja.Promise {
```

Comment:
A lot of code relayed to the promises is duplicating, but the pattern is pretty straightforward.

Do you think we can try to remove this duplicity? :thinking:

Reply:
I've actually considered it while writing the code. But in the end, I really prefer to have the code repeated, as it reduces the cognitive load, and keeps the code really simple and clear: you can look at every function independently without having to go around. Also the repeated code is really small and as I'm not 100% confident around the promises and event loop, I find it easier to manage.

If everyone else disagrees, I'm open to change that, but as of now, I personally would prefer to keep it as is 🙇🏻

Reply:
@mstoykov in fact not really, there is no reflection required, an example:

```go
func (c *Client) Get(key string) *goja.Promise {
	return c.buildPromise(func() (interface{}, error) {
		return c.redisClient.Get(c.vu.Context(), key).Result()
	})
}

func (c *Client) GetSet(key string, value interface{}) *goja.Promise {
	return c.buildPromise(func() (interface{}, error) {
		return c.redisClient.GetSet(c.vu.Context(), key, value).Result()
	}, value)
}

func (c *Client) Del(keys ...string) *goja.Promise {
	return c.buildPromise(func() (interface{}, error) {
		return c.redisClient.Del(c.vu.Context(), keys...).Result()
	})
}

func (c *Client) buildPromise(fn func() (interface{}, error), args ...interface{}) *goja.Promise {
	promise, resolve, reject := c.makeHandledPromise()

	if err := c.isSupportedType(1, args...); err != nil {
		reject(err)
		return promise
	}

	go func() {
		val, err := fn()
		if err != nil {
			reject(err)
		}

		resolve(val)
	}()

	return promise
}
```

@oleiade well, readability is precisely the reason why I raised that discussion (improved maintainability is another benefit). I believe fewer code -> easier to read and learn. That's why from my perspective, it is also separated by concerns is better. But it's subjective. That's why leaving this comment with the example. Maybe that will be easier for to others conclude which approach is better.

Reply:
Hey @oleiade

>Regardless, it's subjective, and a style preference indeed, and I don't think there's an objective better solution in this scenario. I wouldn't want to spend too much energy on that if that's okay with you

It's absolutely okay with me. Sorry if that discussion spends too much energy :cry: It wasn't the point. The point was to provide an alternative opinion as always.

But please don't resolve this thread until the others who can review this PR will check this discussion and also share their opinion :bow: